### PR TITLE
Fix: Allow hashing of metrics with lists in their state

### LINKED ISSUE #5926: AveragePrecision Broken on Master
## 🐛 Bug

On `master` (currently 4857546), the implementation of `__hash__` on PyTorch Lightning metrics causes an exception to be raised for metric implementations which make use of non-hashable state. One such exception is `AveragePrecision`, meaning that any model which uses it is currently broken on `master`.

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1sSMuMoLyeOEtNupTOPC7YuLQonVxBt4H?usp=sharing

### To Reproduce

See above BoringModel. All that's needed to reproduce the bug is installing `master` (currently 4857546) and adding an `AveragePrecision` metric to a model. This exception is raised (copied from above Colab link):

```
/usr/local/lib/python3.6/dist-packages/pytorch_lightning/metrics/metric.py in __hash__(self)
    336             hash_vals.append(getattr(self, key))
    337 
--> 338         return hash(tuple(hash_vals))
    339 
    340     def __add__(self, other: Any):

TypeError: unhashable type: 'list'
```

### Expected behavior

Not crashing 🙂 

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.2.0dev
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by peblair**
## What does this PR do?

Fixes #5926. This PR allows metrics to contain lists within their internal state, so long as the contents of those lists are hashable.

## Before submitting
- [X] Was this discussed/approved via a GitHub issue? #5926 
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [X] Did you make sure to update the documentation with your changes? (if necessary) (N/A)
- [X] Did you write any new necessary tests? (not for typos and docs) (Feel free to push back on this 🙂 )
- [X] Did you verify new and existing tests pass locally with your changes?
- [X] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings) (bug not in released version, so omitted from CHANGELOG)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

Always

### CODE REVIEW

**Comment by SkafteNicki** [comment]
@peblair could you add a test to this file:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/metrics/test_metric.py
that test that metrics are hashable, preferable both metric for which the metric state is a tensor and one where it is a list.

**Review by SeanNaren — APPROVED** [review]
Ran into this issue as well, and this was my fix

**Comment by peblair** [comment]
@SkafteNicki I went to add the test, but I came across something weird in the process. This was my original test:
```python
def test_hash():

    class A(Dummy):
        pass

    class B(DummyList):
        pass

    a1 = A()
    a2 = A()
    assert hash(a1) == hash(a2)

    b1 = B()
    b2 = B()
    assert hash(b1) == hash(b2)
    assert isinstance(b1.x, list) and len(b1.x) == 0
    b1.x.append(torch.tensor(5))
    assert isinstance(hash(b1), int) # <- check that nothing crashes
    assert isinstance(b1.x, list) and len(b1.x) == 1
```
This actually fails on the first `assert`, though, since the `hash()` of each `A` instance's `x` values (both Torch tensors) are different. A quick trip to the CLI confirms this:
```
>>> import torch
>>> t1 = torch.tensor(0.0)
>>> t2 = torch.tensor(0.0)
>>> hash(t1) == hash(t2)
False
```
This is a bit counterintuitive to me, since it appears that the `__eq__` is doing a _structural_ equality check (meaning that their hashes should be the same so long as their contents are the same). I am 90+% sure that this is an even deeper bug in `__hash__` (distinct from the one which I am trying to fix in this PR). How should we proceed?

**Comment by SeanNaren** [comment]
@peblair this might be the reason: https://discuss.pytorch.org/t/why-hash-values-of-a-same-tensors-change-if-we-affect-it-to-different-variables/40028

**Comment by peblair** [comment]
@SeanNaren Understood. The issue is that `__hash__` and `__eq__` are intended to have a specific relationship; specifically, `a == b => hash(a) == hash(b)`. This relationship is important to make data structures such as sets and hash maps behave correctly. This is currently being violated on `master`, though, since `a == b` is doing a structural equality check (i.e. do all of the fields on these metrics have the same _contents_?), whereas `hash(a) == hash(b)` is doing an identity equality check (i.e. are all of the fields on these metrics _the same objects_?).

One of these two implementations needs to be changed; I am unsure which, though, since I am unsure what the original intention was when the `__eq__` and `__hash__` implementations were written. Take the snippet in my above post, for example. Is the intention for `a1 == a2` to be `True`? If so, we need to fix `__hash__`. If not, we need to fix `__eq__`.

**Comment by peblair** [comment]
(To clarify, when I said "This is a bit counterintuitive to me", I specifically mean "This implementation of `__hash__` is a bit counterintuitive to me;" apologies if that led to any confusion)

**Comment by SkafteNicki** [comment]
we should probably get @justusschock opinion
IMO two metrics should be equal if they are the same class and their metric states are equal

**Comment by peblair** [comment]
@SkafteNicki that is my intuition too. Let's see what he has to say, but, if that is the intention, I can propose this alternative implementation of `__hash__` which fixes this PR's bug and the equality issue:
```python
    def __hash__(self):
        hash_vals = [self.__class__.__name__]

        for key in self._defaults.keys():
            val = getattr(self, key)
            # Special case: allow list values, so long as their elements are hashable
            if isinstance(val, list):
                hash_vals.extend(val)
            else:
                hash_vals.append(val)

        # Torch tensors are hashable, but based on their underlying pointer.
        # Since we do a structural equality check below, we circumvent this
        # by hashing the _sum_ of tensors' contents
        hash_vals = [(float(x.detach().cpu().sum()) if isinstance(x, torch.Tensor) else x) for x in hash_vals]

        return hash(tuple(hash_vals))
```
There are performance implications for this, of course, but I think the cost is about as sensible I'd expect any other structurally-based hash to be.

File: pytorch_lightning/metrics/metric.py [hunk 1]
```diff
@@ -333,7 +333,11 @@ def __hash__(self):
         hash_vals = [self.__class__.__name__]
 
         for key in self._defaults.keys():
-            hash_vals.append(getattr(self, key))
+            val = getattr(self, key)
+            # Special case: allow list values, so long as their elements are hashable
+            if isinstance(val, list):
```
Comment (justusschock) [inline pytorch_lightning/metrics/metric.py:338]: are lists the only issue here? should we do it with sequences instead?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/metrics/metric.py (hunk 1 above)
Comment (peblair) [reply pytorch_lightning/metrics/metric.py:338]: This is a good point. I was erring on the side of caution, but perhaps it was too cautious. What about this condition instead?
```
if hasattr(val, '__iter__') and not isinstance(val, torch.Tensor):
```

**Comment by justusschock** [comment]
@peblair I think your suggestion is quite good, but instead 
`hash_vals = [(float(x.detach().cpu().sum()) if isinstance(x, torch.Tensor) else x) for x in hash_vals]`

i'd probably use 
```python
hash_vals = [tuple(x.detach().cpu().tolist()) if isinstance(x, torch.Tensor) else x) for x in hash_vals]
```

since tuples can be hashed, right?

Also
>since a == b is doing a structural equality check (i.e. do all of the fields on these metrics have the same contents?)
should not be true, Instead we actually create a compositional metric there that checks for each output that is compute for `a` and `b` if they're identically. So basically this is kind of a lazy way to check equality

**Comment by peblair** [comment]
> @peblair I think your suggestion is quite good, but instead of
> 
> ```python
> hash_vals = [(float(x.detach().cpu().sum()) if isinstance(x, torch.Tensor) else x) for x in hash_vals]
> ```
> 
> i'd probably use
> 
> ```python
> hash_vals = [tuple(x.detach().cpu().tolist()) if isinstance(x, torch.Tensor) else x) for x in hash_vals]
> ```
> 
> since tuples can be hashed, right?

That is true, but this has the benefit of working for tensors of an arbitrary dimensionality. Moreover, I'd imagine that, under the hood, it's effectively the same amount of work to do `hash(float(sum(tensor)))` as doing `hash(tuple(tensor))` (since both are O(n)...in fact, the summation avoids an extra traversal of the tensor's contents, so it may even be faster).

> Also
> 
> > since a == b is doing a structural equality check (i.e. do all of the fields on these metrics have the same contents?)
> 
> should not be true, Instead we actually create a compositional metric there that checks for each output that is compute for `a` and `b` if they're identically. So basically this is kind of a lazy way to check equality

This muddies things a little bit 🙂 . The defined semantics of these methods are less clear in situations like this, but I would argue that we should err on the side of a structural hash here, since that appears to be a sensible choice (`all(a == b) => hash(a) == hash(b)` doesn't have quite the same ring to it, but it makes sense IMO). So, should I move forward with a version of the proposed solution in my last post?

**Comment by peblair** [comment]
I have gone ahead and pushed the discussed changes. I am happy to discuss any further feedback, but hopefully this addresses everything. Thanks for being so responsive, everyone!

File: pytorch_lightning/metrics/metric.py [hunk 2]
```diff
@@ -332,8 +332,23 @@ def _filter_kwargs(self, **kwargs):
     def __hash__(self):
         hash_vals = [self.__class__.__name__]
 
+        # Torch tensors are hashable, but based on their
+        # underlying pointer. Since we do a structural
+        # equality check in __eq__, we circumvent this
+        # by hashing the _sum_ of tensors' contents
+        def format_tensor(t):
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:339]: Please, no local functions, make it a staticmethod
add the above comments as a docstring to the function.
```suggestion
        def hash_tensor(t):
```

**Review by SkafteNicki — APPROVED** [review]
single comment, else LGTM

File: pytorch_lightning/metrics/metric.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/metrics/metric.py:339]: Let's make this function private `_hash_tensor`

File: tests/metrics/test_metric.py [hunk 3]
```diff
@@ -154,6 +154,27 @@ def compute(self):
     assert a.compute() == 5
 
 
+def test_hash():
```
Comment (tchaton) [inline tests/metrics/test_metric.py:157]: Could this test be done directly on Metric objects ?

**Review by tchaton — COMMENTED** [review]
Looks good, small nits.

**Comment by Borda** [comment]
@peblair mind check the last remaining comments above?

File: tests/metrics/test_metric.py (hunk 3 above)
Comment (peblair) [reply tests/metrics/test_metric.py:157]: @tchaton I apologize, could you elaborate? The hashing is being done on two subclasses of `Metric`, so I am unsure what precisely you mean.

**Comment by peblair** [comment]
I unfortunately discovered a major issue with this change today, which I think means that we need to revisit the semantics issue. You can see a reproduction on the `BoringModule` [here](https://colab.research.google.com/drive/1zNCvLu6rQlY4Z2Lgz9g3I3l91gCG7lYq?usp=sharing), but here is the explanation: PyTorch's `nn.Module`'s `children()` method [uses a `set` under the hood](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1347-L1365) to memoize the parameters on modules. This means that, if you have two metrics which hash the same and are `__eq__` to one another, then `torch.nn.Module.named_children()` (and, by extension, `.children()`) **will only return one of them**. This is the case when the metrics' contents are the same, such as during initialization. You can see in the Colab link that this causes issues; because `torch.nn.Module.to()` delegates out to `children()`, if you have, say, two MSE metrics on a `LightningModule`, **only one gets moved to the GPU**.

So, what to do? Here are some possible options:
1. Revert to the original hashing semantics (plus the list fix which originally inspired this PR). I still think this is counterintuitive, but it won't crash.
2. Same as no. 1, but also modify the semantics of `__eq__` to better reflect the `__hash__` semantics (e.g. doing an identity equality check on underlying tensors).
3. Something else that someone more clever than I comes up with 🙂 

cc @SkafteNicki @justusschock

**Comment by SkafteNicki** [comment]
Then I am in favour of option 1, because else `__eq__` will behave differently than other logical operators (I was originally in favour of structual equality, but then I think it should be the same for all logical operators).
That said we should probably clarify this in the documentation, as we right now only have an example of adding two metrics (which makes sense) but not one of a logical operator.

**Comment by peblair** [comment]
Reflecting on this a bit more, I am wondering if the semantics of `__eq__` should be changed. The fact that two tensors hash to different values is mostly a coincidence; I can imagine that this will lead to some confusing errors for people who implement their own metrics (imagine that you had implemented the MSE metric yourself and randomly only one of your metrics was being placed on the GPU)...at a minimum, I think the semantics should be clearly defined and documented.

**Comment by Borda** [comment]
@peblair mind check the failing test on GPU

**Comment by justusschock** [comment]
I am also in favor of of number one (mostly for the same reasons as @SkafteNicki)

However, we definitely need to fix the moving issue in a separate PR (not yet sure how though)

**Comment by SkafteNicki** [comment]
I think the way forward is to keep the implementation as it is now. It fixes the current bug without introducing a new one (the moving issue is only a problem if we try change the hashing function to do structual similarity).
The tests will have to be changed to
``` python
a1 = A()
a2 = A()
assert hash(a1) != hash(a2)
```
so the tests document that we are **not** expecting two metric objects to be the same even if their content is the same (basically following pytorch standards). 

We can do a follow up PR where we document this behaviour better (that evaluating if two metrics are equal does not check if it is the same metric, but if the output from their `compute` is the same).

**Comment by peblair** [comment]
The problematic test is now fixed to be consistent with the new semantics, but I think that the situation is still precarious (on `master`, before this change). Because `__eq__` returns a non-`bool`, this line passes if I add it to `test_hash`:
```python
assert (b1 == b2) and (b1 != b2)
```
Now, that's _fine_, I suppose, except for the fact that I am 99% sure that `torch.nn.Module.named_children()` is now _only_ not broken because different metric instances return different hashes, so they _happen_ to be able to both be added to the underlying `set()` that `named_children` is using (since they have different hashes, they end up in different buckets in the hash set implementation CPython uses). I feel like this is asking for nondeterministic failures to crop up, since a bad pair of pointers could both hash (mod `python_dict_size`) to the same bucket. I'm wondering if the solution here (which is beyond the scope of this PR at this point) is to escalate to PyTorch. Maybe their memoization in `named_children()` should contain the `id()` of each object so as to only check for identity equality...

**Pull request merged**